pkg/network/node/iptables.go

```diff
@@ -227,7 +227,7 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
 
 func (n *NodeIPTables) ensureEgressIPRules(egressIP, mark string) error {
 	for _, cidr := range n.clusterNetworkCIDR {
-		_, err := n.ipt.EnsureRule(iptables.Prepend, iptables.TableNAT, iptables.Chain("OPENSHIFT-MASQUERADE"), "-s", cidr, "-m", "mark", "--mark", mark, "-j", "SNAT", "--to-source", egressIP)
+		_, err := n.ipt.EnsureRule(iptables.Prepend, iptables.TableNAT, iptables.Chain("OPENSHIFT-MASQUERADE"), "-s", cidr, "!", "-d", cidr, "-m", "mark", "--mark", mark, "-j", "SNAT", "--to-source", egressIP)
```

Comment:
Hmm, this does fix the bug but I am not sure it's the proper way to fix it. We shouldn't be sending packets to egress nodes if those packets are east <-> west related (which seems to be the case, given that this bug exists). I think the proper way to fix it would be to add a default OVS flow in `SetupOVS` in table 100 that takes precedence over the egress IP related flows that are later setup, similar to what is done here: https://github.com/openshift/sdn/blob/master/pkg/network/node/ovscontroller.go#L203-L206

As such we'll avoid additional hops to egress node for packets that are not actually "egress-ing" the cluster.

Reply:
@alexanderConstantinescu you can find more background in the bugzilla referenced in the title of the PR.

I am not sure what you mean by "egress nodes". Egress IP is implemented by S-NAT in iptables on all the nodes.

First the packets have the load balancer IP as destination, outside of the pod range and as such it is not identified in OVS that it is east <-> west traffic. It is only after the D-NAT is applied in the load balancer implementation that it is clear that it is east <-> west traffic. This is again implemented in iptables not in OVS. This PR is only aiming at fixing the issue so that egress IP plays well with services of type load balancer not to change the implementation strategy of one or the other.

With the proposed fix the packets won't be sent to a different node outside of the VXLAN. I am wondering whether you are not thinking of egress router rather than egress IP. The issue here is with the later.

Reply:
Yes, reading more closely I realize that it's true that on clouds such as Azure, pods will send their traffic to an external VIP when targeting an OpenShift route, which thus means that we can't implement what I said in OVS.   

> Egress IP is implemented by S-NAT in iptables on all the nodes.

No, that is wrong. There is a notion of "nodes selected by the user to be used as egress node for egress IP", this is what I referred to as "egress nodes" - and this is done by the user by specifying `hostsubnet.spec.egressIPs` or `hostsubnet.spec.egressCIDRs` on the node they would like to act as "egress nodes". The iptables rules you've modified are only applied on those nodes, not on all cluster nodes. 

Ideally, traffic for OpenShift routes should not even be routed to these egress nodes, since the endpoint is an internal cluster pod. However, as we've just discussed: that is not possible since the cloud's virtual IP space is not identifiable by openshift-sdn (at least today) and we can't tell OVS to ignore that traffic.

Reply:
That being said I am however thinking the customer might have done something wrong when claiming the egress IP on Azure in their setup. If they've done that right I don't see why Azure would not be able to route the egress IP correctly? However, your patch seems fine for what concerns the added filter in the iptables rules used.

Reply:
> That being said I am however thinking the customer might have done something wrong when claiming the egress IP on Azure in their setup. If they've done that right I don't see why Azure would not be able to route the egress IP correctly? 

There is more information in the bugzilla but the traffic does not leave the node with the egress IP as far as I have seen. I was also able to reproduce on a newly installed cluster. Here is what I understand from the flow:

PREROUTING

Chain KUBE-SEP-GMF6HD62GMXKGYYH (1 references)
target     prot opt source               destination
KUBE-MARK-MASQ  all  --  10.130.4.33          0.0.0.0/0            /* openshift-ingress/router-default:http */
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            /* openshift-ingress/router-default:http */ tcp to:10.130.4.33:80

At this point the destination IP has been changed from the load balancer IP to the IP of a router pod.

POSTROUTING

Chain OPENSHIFT-MASQUERADE (1 references)
target     prot opt source               destination
SNAT       all  --  10.128.0.0/14        0.0.0.0/0            mark match 0x6e52f2 to:10.0.116.8

Masquerading is done on all pods with the egress IP for packets that have a mark matching 0x6e52f2
At this point we have the packets that have the egress IP as source IP and router pod IP as destination IP. This should go to OVS but OVS will not accept traffic from external IP (the egress IP is not the IP of the node) to get directly routed to the pod and the packets get dropped.


> However, your patch seems fine for what concerns the added filter in the iptables rules used.

Good to read. I am looking forward to getting this progressed.

Reply:
@danwinship might have some input here too